### ISSUE TITLE
About CUDA Memory overflow, halc_dola_beam_search, n=2m?, valid list = ["NOUN", "PROPN", "ADD"]

### ISSUE THREAD

**Issue by JiHoonLee9898** [BillChan226/contrast_decoding_LVLMs#20]
Hello,

First of all, I would like to express my gratitude for sharing this remarkable research and an excellent codebase. I truly appreciate the effort and transparency behind this work.
I have a few questions regarding an experiment I am conducting. Specifically, I am attempting to increase the maximum number of generated tokens from 64 and verify the chair score at each step, as shown in Figure 4 of the paper (64, 128, 192, 256, 320, 384).

1.
**Could you advise on the GPU resources required for such an experiment? Would multiple A6000 GPUs(*4) be sufficient, or would this be infeasible?**

When using the default settings specified in the README on an NVIDIA RTX A6000 (vRAM 49140MiB) but increasing the sample count to 500, I encounter a CUDA memory overflow. The default settings limit the generated token count to 64, but I aim to extend this to 384, as in Figure 4 of the paper, and measure the scores accordingly.
I’ve noticed that setting multiple GPUs as visible devices causes tensors to be distributed across different GPUs. If I use a single A100 GPU, would it be adequate to perform the experiment for 384 tokens as shown in Figure 4?

2.
**When halc_dola_beam_search is invoked in decode mode = halc, is this intentional and in alignment with the methodology proposed in the paper?** If my understanding is incorrect, please let me know. Should I replace this function with another in transformers-4.36.2/src/transformers/generation/utils.py (e.g., halc_greedy_decode) that includes "halc" in its name? If so, which function would be the correct one?

When the decoding mode is set to halc, captions are generated using the halc_dola_beam_search function in HALC/transformers-4.36.2/src/transformers/generation/utils.py. (For instance, in the case of LLaVA, the workflow seems to be as follows: the generate() function in transformers-4.36.2/utils.py returns ids, the generate() function in minigpt4/models/llava.py converts ids to text, and finally, model.generate() returns the text.)
Upon reviewing the function, it seems that when a word comprises two or more tokens (e.g., _giraffe), contrastive decoding (not greedy decoding) is used to generate tokens like _gi, raf, and fe. This approach seems slightly different from the ratio-based method of DOLA(appearing more like subtracting alpha from 1+alpha). However, I could not find a reference to this in the paper.

3.
In the context_layer_double_multi_contrastive_decoding function in halc.py, it appears that the logic selects 2m candidates per beam, resulting in 2mk total candidates. While the paper does not seem to explicitly state this, **the code appears to assume n=2m.** This suggests that the focal view must always be even, and among the focal view Combination 2 candidates, the top m are selected to produce 2m contrast logits. **Is my understanding correct?**

4.
**Should I modify the valid_list to reproduce the results in the paper?**
In HALC/decoder_zoo/HALC/context_density/halc.py, the valid list is defined as ["NOUN", "PROPN", "ADD"]. 
However, the paper states that the process applies to tokens of other POS tags:
“If the token belongs to noun, adjective/adverb/number/verb/pronoun, or preposition, which correspond to object existence, attribute, and relationship hallucinations, respectively, we redo the current token generation with HALC…” (page 3).
Considering this discrepancy, should I modify the valid_list to reproduce the results in the paper?

Thank you for your time and assistance. I greatly appreciate your insights on these matters.

**Comment by sev777** [comment]
‘ the sample count to 500’ is the args.num_samples=500? But in the code
`    # process segmentation ground truth
    processed_segment_results = []
    # Sample images which contain more than sample_num objects
    for cur_image in segment_results:
        if len(cur_image["objects"]) >= num_samples:
            processed_segment_results.append(cur_image)
`
the 500 means one image have 500+ object? which is strange。
